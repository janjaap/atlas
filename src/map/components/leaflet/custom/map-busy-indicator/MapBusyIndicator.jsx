// inspired from https://github.com/ebrelsford/Leaflet.loading
// and https://github.com/tumerorkun/react-leaflet-zoom-indicator
import L from 'leaflet';
import { MapControl } from 'react-leaflet';
import './MapBusyIndicator.scss';

export const DATA_LOADING_EVENT = 'dataloading';
export const DATA_LOADED_EVENT = 'dataloaded';

export default class MapBusyIndicator extends MapControl {

  constructor(props, context) {
    super(props);
    this.map = context.map;
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.loading !== nextProps.loading) {
      if (nextProps.loading) {
        this.map.fire(DATA_LOADING_EVENT);
      } else {
        this.map.fire(DATA_LOADED_EVENT);
      }
    }
  }

  shouldComponentUpdate() {
    return false;
  }

  createLeafletElement(props) { // eslint-disable-line class-methods-use-this
    const MapBusyIndicatorControl = this.createMapBusyIndicatorControl();
    return new MapBusyIndicatorControl(props);
  }

  createMapBusyIndicatorControl() { // eslint-disable-line class-methods-use-this
    const controlExtension = this.createMapContolExtension();
    return L.Control.extend(controlExtension);
  }

  createMapContolExtension() { // eslint-disable-line class-methods-use-this
    return {
      options: {
        position: 'topright'
      },

      initialize(options) {
        L.setOptions(this, options);
        this.state = {
          pendingCalls: 0
        };
      },

      createContainer() {
        //  create a container for the indicator
        this.indicatorContainer = L.DomUtil.create('div', 'leaflet-control-layer-container leaflet-bar');
        this.indicatorContainer.innerHTML = `
          <div class="c-loading-indicator c-loading-indicator--box">
              <img class="c-loading-indicator__icon" src="assets/images/spinner.svg" alt="">
              <span class="c-loading-indicator__text">Bezig met laden...</span>
          </div>
        `;
        return this.indicatorContainer;
      },

      onAdd(map) {
        this.addMapListeners(map);
        const container = this.createContainer();
        return container;
      },

      onRemove(map) {
        this.removeMapListeners(map);
      },

      updateIndicator(isLoading) {
        const oldState = this.state;
        if (isLoading) {
          this.state = {
            pendingCalls: this.state.pendingCalls + 1
          };
        } else {
          this.state = {
            pendingCalls: this.state.pendingCalls > 0 ? this.state.pendingCalls - 1 : 0
          };
        }
        if (oldState.pendingCalls === 0 && this.state.pendingCalls > 0) {
          L.DomUtil.addClass(this.indicatorContainer, 'is-loading');
        } else if (this.state.pendingCalls === 0) {
          L.DomUtil.removeClass(this.indicatorContainer, 'is-loading');
        }
      },

      handleLoading() {
        this.updateIndicator(true);
      },

      handleLoaded() {
        this.updateIndicator(false);
      },

      addMapListeners(map) {
        // Add listeners to the map for (custom) DATA_LOADING_EVENT and DATA_LOADED_EVENT
        // events, for AJAX calls that affect the map but will not be
        // reflected in the layer events.
        map.on({
          [DATA_LOADING_EVENT]: this.handleLoading,
          [DATA_LOADED_EVENT]: this.handleLoaded
        }, this);
      },

      removeMapListeners(map) {
        map.off({
          [DATA_LOADING_EVENT]: this.handleLoading,
          [DATA_LOADED_EVENT]: this.handleLoaded
        }, this);
      }
    };
  }
}
