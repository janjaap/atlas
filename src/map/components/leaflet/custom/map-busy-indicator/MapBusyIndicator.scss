@import "../../../../../shared/styles/config";

.leaflet-control.leaflet-bar.leaflet-control-layer-container { // sass-lint:disable-line force-element-nesting
  border: none;
  display: none;
  margin-right: 0;
  margin-top: $base-whitespace - 2px;
  right: $base-whitespace * 5;

  &.is-loading {
    display: block;
  }
}

.leaflet-container {
  .c-loading-indicator {
    font-family: $base-font-family;
    font-size: $s-font-size;
    height: 3 * $base-whitespace + 3px;
    left: 0;
    position: relative;
    top: 0;
  }
}
