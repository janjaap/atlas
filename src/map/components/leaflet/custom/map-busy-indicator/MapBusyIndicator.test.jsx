import React from 'react';
import L from 'leaflet';
import { shallow } from 'enzyme';
import MapBusyIndicator, { DATA_LOADING_EVENT, DATA_LOADED_EVENT } from './MapBusyIndicator';

jest.mock('leaflet');
jest.mock('react-leaflet');

describe('MapBusyIndicator', () => {
  let wrapper;
  let wrapperInstance;
  const map = {
    on: jest.fn(),
    off: jest.fn(),
    fire: jest.fn()
  };


  beforeAll(() => {
    L.Control.extend.mockImplementation(() => jest.fn());
    L.DomUtil.create.mockImplementation(() => ({}));
  });

  beforeEach(() => {
    wrapper = shallow(
      <MapBusyIndicator
        loading={false}
      />, {
        context: {
          map
        }
      }
    );
    wrapperInstance = wrapper.instance();
  });

  it('should render an empty component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should add the control to the leaflet instance', () => {
    wrapperInstance.createLeafletElement();
    expect(L.Control.extend).toHaveBeenCalled();
  });

  it('should never update', () => {
    expect(wrapperInstance.shouldComponentUpdate()).toBe(false);
  });

  it('should show the indicator when the loading prop is set', () => {
    wrapperInstance.componentWillReceiveProps({ loading: true });
    expect(map.fire).toHaveBeenCalledWith(DATA_LOADING_EVENT);
  });

  it('should do nothing when the loading prop is not changed', () => {
    map.fire.mockClear();
    expect(map.fire).not.toHaveBeenCalled();
    wrapper.setProps({ loading: false });
    wrapperInstance.componentWillReceiveProps({ loading: false });
    expect(map.fire).not.toHaveBeenCalled();
  });

  it('should hide the indicator when the loading prop is set', () => {
    wrapper.setProps({ loading: true });
    wrapperInstance.componentWillReceiveProps({ loading: false });
    expect(map.fire).toHaveBeenCalledWith(DATA_LOADED_EVENT);
  });

  describe('mapControlExtension', () => {
    let mapControlExtension;
    beforeEach(() => {
      mapControlExtension = wrapperInstance.createMapContolExtension();
      L.DomUtil.addClass.mockClear();
      L.DomUtil.removeClass.mockClear();
    });

    it('should create mapControlExtension', () => {
      expect(mapControlExtension).toBeDefined();
    });

    it('should inititalize the control', () => {
      const options = { foo: 'bar' };
      mapControlExtension.initialize(options);
      expect(L.setOptions).toHaveBeenLastCalledWith(mapControlExtension, options);
      expect(mapControlExtension.state).toEqual({ pendingCalls: 0 });
    });

    it('should hook the loading events and add the control to the map', () => {
      mapControlExtension.onAdd(map);
      expect(map.on).toHaveBeenCalled();
    });

    it('should remove the events when the map is unloaded', () => {
      mapControlExtension.onRemove(map);

      expect(map.off).toHaveBeenCalled();
    });

    it('should handleLoading when no call is pending', () => {
      mapControlExtension.state = { pendingCalls: 0 };
      mapControlExtension.handleLoading();
      expect(L.DomUtil.addClass).toHaveBeenCalled();
      expect(mapControlExtension.state).toEqual({ pendingCalls: 1 });
    });

    it('should handleLoading when calls are pending', () => {
      mapControlExtension.state = { pendingCalls: 2 };
      mapControlExtension.handleLoading();
      expect(L.DomUtil.addClass).not.toHaveBeenCalled();
      expect(L.DomUtil.removeClass).not.toHaveBeenCalled();
      expect(mapControlExtension.state).toEqual({ pendingCalls: 3 });
    });

    it('should handleLoaded when one call is pending', () => {
      mapControlExtension.state = { pendingCalls: 1 };
      mapControlExtension.handleLoaded();
      expect(L.DomUtil.removeClass).toHaveBeenCalled();
      expect(mapControlExtension.state).toEqual({ pendingCalls: 0 });
    });

    it('should handleLoaded when more then one call is pending', () => {
      mapControlExtension.state = { pendingCalls: 3 };
      mapControlExtension.handleLoaded();
      expect(L.DomUtil.removeClass).not.toHaveBeenCalled();
      expect(mapControlExtension.state).toEqual({ pendingCalls: 2 });
    });
  });
});
