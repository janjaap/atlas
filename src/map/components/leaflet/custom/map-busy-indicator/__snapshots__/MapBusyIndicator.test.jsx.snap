// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBusyIndicator should render an empty component 1`] = `""`;
